pkgs/development/libraries/kerberos/heimdal.nix

```diff
@@ -1,65 +1,119 @@
-{ lib, stdenv, fetchFromGitHub, autoreconfHook, pkg-config, python3, perl, bison, flex
-, texinfo, perlPackages
-, openldap, libcap_ng, sqlite, openssl, db, libedit, pam
-, CoreFoundation, Security, SystemConfiguration
+{ lib
+, stdenv
+, fetchFromGitHub
+, autoreconfHook
+, pkg-config
+, python3
+, perl
+, bison
+, flex
+, texinfo
+, perlPackages
+
+, openldap
+, libcap_ng
+, sqlite
+, openssl
+, db
+, libedit
+, pam
+, krb5
+, libmicrohttpd
+, cjson
+
+, CoreFoundation
+, Security
+, SystemConfiguration
+
+, curl
+, jdk
+, unzip
+, which
+
+, nixosTests
+
+, withCJSON ? true
+, withCapNG ? stdenv.isLinux
+, withMicroHTTPD ? true
+, withOpenLDAP ? true
+, withOpenLDAPAsHDBModule ? false
+, withOpenSSL ? true
+, withSQLite3 ? true
 }:
 
-stdenv.mkDerivation rec {
+assert lib.assertMsg (withOpenLDAPAsHDBModule -> withOpenLDAP) ''
+  OpenLDAP needs to be enabled in order to build the OpenLDAP HDB Module.
+'';
+
+stdenv.mkDerivation {
   pname = "heimdal";
-  version = "7.8.0";
+  version = "7.8.0-unstable-2023-11-29";
 
   src = fetchFromGitHub {
     owner = "heimdal";
     repo = "heimdal";
-    rev = "heimdal-${version}";
-    sha256 = "sha256-iXOaar1S3y0xHdL0S+vS0uxoFQjy43kABxqE+KEhxjU=";
+    rev = "3253c49544eacb33d5ad2f6f919b0696e5aab794";
+    hash = "sha256-uljzQBzXrZCZjcIWfioqHN8YsbUUNy14Vo+A3vZIXzM=";
   };
 
   outputs = [ "out" "dev" "man" "info" ];
 
-  patches = [ ./heimdal-make-missing-headers.patch ];
-
-  nativeBuildInputs = [ autoreconfHook pkg-config python3 perl bison flex texinfo ]
-    ++ (with perlPackages; [ JSON ]);
-  buildInputs = lib.optionals (stdenv.isLinux) [ libcap_ng ]
-    ++ [ db sqlite openssl libedit openldap pam]
-    ++ lib.optionals (stdenv.isDarwin) [ CoreFoundation Security SystemConfiguration ];
+  nativeBuildInputs = [
+    autoreconfHook
+    pkg-config
+    python3
+    perl
+    bison
+    flex
+    texinfo
+  ]
+  ++ (with perlPackages; [ JSON ]);
+
+  buildInputs = [ db libedit pam ]
+    ++ lib.optionals (stdenv.isDarwin) [ CoreFoundation Security SystemConfiguration ]
+    ++ lib.optionals (withCJSON) [ cjson ]
+    ++ lib.optionals (withCapNG) [ libcap_ng ]
+    ++ lib.optionals (withMicroHTTPD) [ libmicrohttpd ]
+    ++ lib.optionals (withOpenLDAP) [ openldap ]
+    ++ lib.optionals (withOpenSSL) [ openssl ]
+    ++ lib.optionals (withSQLite3) [ sqlite ];
+
+  doCheck = true;
+  nativeCheckInputs = [
+    curl
+    jdk
+    unzip
+    which
+  ];
 
-  ## ugly, X should be made an option
   configureFlags = [
     "--sysconfdir=/etc"
     "--localstatedir=/var"
```

Comment:
I feel like these two also shouldn't be needed since they are the standard directories. Not sure though.

Reply:
`configure.ac` seems to not specify defaults but remove the `/usr/heimdal` prefix, if it is set? Not confident I'm reading this correctly

https://github.com/heimdal/heimdal/blob/3253c49544eacb33d5ad2f6f919b0696e5aab794/configure.ac#L27-L30

Reply:
Ah yeah, it uses /etc and /var/heimdal for these if unset. IIRC /var/heimdal is used for the KDC? Probably want to at least keep that one then so this doesn't break existing KDC setups (though, not sure there are any considering this package has been broken for a while :P).

Looking at the strings in the binaries, these are the relevant path refs I can find:

```
FILE:/var/heimdal/user/%{euid}/client.keytab
/var/heimdal
/var/heimdal/heimdal
/var/heimdal/kadmind.acl
/var/heimdal/m-key
```

With "--localstatedir=/var", the only difference seems to be that the first line is now

```
FILE:/var/user/%{euid}/client.keytab
```

Of course, that doesn't have to mean anything considering there can be more paths concated.

So it might be a good idea to change that regardless since at the very least /var/user is kind of a bad place to put stuff since you don't know what it belongs to at first glance.

Reply:
I'm not able to reproduce the `/var/user` thing. I've tried grepping through `result` with `rg --binary "/var/user"`. Do you mind writing up the commands you used to find these?

Reply:
I'll check in a few days, I only have a Mac right now and it doesn't compile there (posted the log in another comment)

Reply:
> I'm not able to reproduce the `/var/user` thing. I've tried grepping through `result` with `rg --binary "/var/user"`. Do you mind writing up the commands you used to find these?

```
[nix-shell:~/.cache/nixpkgs-review/pr-273807/results/heimdal]$ strings lib/libkrb5.so | grep /var/user
FILE:/var/user/%{euid}/client.keytab
```

Reply:
This seems like they are saying these are the defaults: https://github.com/heimdal/heimdal/blob/cb9a13032285ee9a20507f21c44b67a547a9996c/ChangeLog.2002#L383-L389. I have removed the flags.